### ISSUE TITLE
Channel based producer does not write to the given partition

### ISSUE THREAD

**Issue by sapei** [confluentinc/confluent-kafka-go#112]
Description
===========
I would like to use channel based producer to publish messages to a specific kafka partition. For this purpose, I set the partition in every kafka message. However the messages won't be delivered to specific partition. They will be instead delivered to any partitions.

How to reproduce
================

Use the [examples/producer_channel_example](examples/producer_channel_example) and set a specific partition in the kafka message on the following line:
`p.ProduceChannel() <- &kafka.Message{TopicPartition: kafka.TopicPartition{Topic: &topic, Partition: int32(0)}, Value: []byte(value)}`

Checklist
=========
Please provide the following information:

 - [ x ] confluent-kafka-go and librdkafka version (`LibraryVersion(0.11.3)`):
 - [ ] Apache Kafka broker version:
 - [ ] Client configuration: `ConfigMap{...}`
 - [ ] Operating system:
 - [ ] Provide client logs (with `"debug": ".."` as necessary)
 - [ ] Provide broker log excerpts
 - [ ] Critical issue

**Comment by edenhill** [comment]
I can't reproduce this,
What Partition are you getting back in the delivery report?

**Comment by sapei** [comment]
Maybe the following snippet is useful to reproduce the problem which I have described before:

`       p, err := kafka.NewProducer(&kafka.ConfigMap{
		"bootstrap.servers": broker,
		"go.batch.producer":   true,
	})
.`
`	p.ProduceChannel() <- &kafka.Message{TopicPartition: kafka.TopicPartition{Topic: &topic, Partition: int32(0)}, Value: []byte(value)}`

**Comment by derrickburns** [comment]
I have observed this bug as well.

**Comment by derrickburns** [comment]
Is this line correct?

https://github.com/confluentinc/confluent-kafka-go/blob/master/kafka/producer.go#L199

**Comment by derrickburns** [comment]
I believe that the use of `C.RD_KAFKA_PARTITION_UA` is the problem.

```
func (p *Producer) produceBatch(topic string, msgs []*Message, msgFlags int) error {
	crkt := p.handle.getRkt(topic)

	cmsgs := make([]C.rd_kafka_message_t, len(msgs))
	for i, m := range msgs {
		p.handle.messageToC(m, &cmsgs[i])
	}
	r := C.rd_kafka_produce_batch(crkt, C.RD_KAFKA_PARTITION_UA, C.int(msgFlags)|C.RD_KAFKA_MSG_F_FREE,
		(*C.rd_kafka_message_t)(&cmsgs[0]), C.int(len(msgs)))
	if r == -1 {
		return newError(C.rd_kafka_last_error())
	}

	return nil
}
```

**Comment by edenhill** [comment]
This is actually a problem with the underlying librdkafka rd_kafka_produce_batch() API that won't honour per-message partitions:
https://github.com/edenhill/librdkafka/blob/master/src/rdkafka.h#L2735

We would need to alter that API to use the per-message partition, without breaking existing applications.

Thanks for the digging!

**Comment by rikardev** [comment]
Yes I think this is still true when using the batch producer. Wrote this code to test and found the correct config value that breaks the manual partition assignment:

```
package main

import (
	"github.com/confluentinc/confluent-kafka-go/kafka"
)

func main() {

	config := kafka.ConfigMap{
		"bootstrap.servers": "localhost:9092",
		"go.batch.producer": true,
	}

	p, _ := kafka.NewProducer(&config)

	topic := "test-topic"

	p.ProduceChannel() <- &kafka.Message{
		TopicPartition: kafka.TopicPartition{
			Topic:     &(topic),
			Partition: 12, // 50 partitions on test-topic
		},
		Key:   []byte("testkey"),
		Value: []byte("testvalue"),
	}

	p.Close()
}
```

This is what I get:
Key:         testkey
Partition:   40
Offset:      1
Timestamp:   2019-10-04 16:49:28.104 +0200 CEST
testvalue

**Comment by edenhill** [comment]
I believe the call to `rd_kafka_produce_batch()` [here](https://github.com/confluentinc/confluent-kafka-go/blob/master/kafka/producer.go#L299) is missing the `RD_KAFKA_MSG_F_PARTITION` flag to honour the specified partition.
